### ISSUE TITLE
"Getting code actions" holds up saving for too long

### ISSUE THREAD

**Issue by camillol** [microsoft/vscode#209789]
Type: <b>Bug</b>

- Have a few extensions and a big Python workspace.
- Enable PyLance type checking.
- Work, work...
- Maybe have a big file search in progress.
- Try to save a file.

You often get a window saying "Saving 'foo.py': Getting code actions from (a bunch of extensions)". This can take many seconds. Sometimes it just gets stuck and you have to reload the window.

<img width="458" alt="Screenshot 2024-04-07 at 7 06 10 PM" src="https://github.com/microsoft/vscode/assets/196272/63713f3e-d3b5-4d63-a72b-e2fa11d3fb05">

My impression is that this high-priority user action is getting stuck behind lower-priority tasks (e.g. reindexing a file, doing a large search), possibly even in another window. When I'm saving a file, if the save actions need e.g. the PyLance LSP, they should preempt other pending tasks, or at least run in parallel (with higher priority).

Secondly, the only option should not be to cancel saving. At least give me a button to save the file as-is, skipping the on-save actions.

VS Code version: Code - Insiders 1.89.0-insider (Universal) (d6eba9b861e3ab7d1935cff61c3943e319f5c830, 2024-04-05T05:53:04.895Z)
OS version: Darwin arm64 23.4.0
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Apple M2 Max (12 x 24)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|6, 7, 6|
|Memory (System)|32.00GB (0.06GB free)|
|Process Argv|--crash-reporter-id 7b87a5b2-68e7-4d42-ad22-32e3bce2e5e7|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (22)</summary>

Extension|Author (truncated)|Version
---|---|---
ruff|cha|2024.16.0
gitlens|eam|14.9.0
copilot|Git|1.178.0
copilot-chat|Git|0.14.0
vscode-pull-request-github|Git|0.86.0
go|gol|0.41.2
vscode-cython|ktn|1.0.3
marp-vscode|mar|2.8.0
python-path|mge|0.0.14
black-formatter|ms-|2024.2.0
debugpy|ms-|2024.5.10961012
isort|ms-|2023.13.10681007
python|ms-|2024.4.0
vscode-pylance|ms-|2024.4.1
jupyter|ms-|2024.3.0
jupyter-keymap|ms-|1.1.2
jupyter-renderers|ms-|1.0.17
vscode-jupyter-cell-tags|ms-|0.1.9
vscode-jupyter-slideshow|ms-|0.1.6
makefile-tools|ms-|0.9.10
sourcegraph|sou|2.2.16
even-better-toml|tam|0.19.2

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vspor879:30202332
vspor708:30202333
vspor363:30204092
vstes627:30244334
vscod805:30301674
vsaa593cf:30376535
py29gd2263:30784851
c4g48928:30535728
vscrpc:30624061
962ge761:30841072
pythongtdpath:30726887
welcomedialog:30812478
pythonidxpt:30768918
pythonnoceb:30776497
asynctok:30898717
dsvsc013:30777762
dsvsc014:30777825
dsvsc015:30821418
pythontestfixt:30866404
pythonregdiag2:30926734
pyreplss1:30879911
pythonmypyd1:30859725
pythoncet0:30859736
2e7ec940:31000449
pythontbext0:30879054
accentitlementst:30870582
dsvsc016:30879898
dsvsc017:30880771
dsvsc018:30880772
bf62j303:30968145
cppperfnew:30980852
8082a590:30961435
fegfb526:30952798
bg6jg535:30979844
ccp2r3:30958153
pythonait:30973460
gee8j676:31009450
gd77d436:30999572

```

</details>

**Comment by mbmccoy** [comment]
Seems related to my issue that was closed (prematurely, IMO) [here](https://github.com/microsoft/vscode/issues/204648).

**Comment by mbmccoy** [comment]
Also related on the ruff page: https://github.com/astral-sh/ruff-vscode/issues/428

**Comment by camillol** [comment]
This has gotten significantly worse for me in the latest Insiders.

Version info:
```
Version: 1.89.0-insider (Universal)
Commit: 50a6f4f200c5a33f47997eb6a8966e50cf219e21
Date: 2024-04-12T05:52:00.589Z (2 days ago)
Electron: 28.2.8
ElectronBuildId: 27744544
Chromium: 120.0.6099.291
Node.js: 18.18.2
V8: 12.0.267.19-electron.0
OS: Darwin arm64 23.4.0
```

A/B tests:
```
vsliv368:30146709
vspor879:30202332
vspor708:30202333
vspor363:30204092
vstes627:30244334
vscod805:30301674
vsaa593cf:30376535
py29gd2263:30784851
c4g48928:30535728
vscrpc:30624061
962ge761:30841072
pythongtdpath:30726887
welcomedialog:30812478
pythonidxpt:30768918
pythonnoceb:30776497
asynctok:30898717
dsvsc013:30777762
dsvsc014:30777825
dsvsc015:30821418
pythontestfixt:30866404
pythonregdiag2:30926734
pyreplss1:30879911
pythonmypyd1:30859725
pythoncet0:30859736
2e7ec940:31000449
pythontbext0:30879054
accentitlementst:30870582
dsvsc016:30879898
dsvsc017:30880771
dsvsc018:30880772
bf62j303:30968145
cppperfnew:30980852
8082a590:30961435
fegfb526:30952798
bg6jg535:30979844
ccp2r3:30958153
pythonait:30973460
gee8j676:31009450
showvideot:31016890
chatpanelt:31014475
gd77d436:30999572
9c06g630:31013171
```

**Comment by camillol** [comment]
One thing that helped was removing the `"source.fixAll": "explicit"` I had for `[python]`.

**Comment by jrieken** [comment]
> My impression is that this high-priority user action is getting stuck behind lower-priority tasks (e.g. reindexing a file, doing a large search), possibly even in another window.

Unsure how we can make that more obvious but it is not low-prio tasks but extensions that do wanna participant in saving that file. That's why saving is waiting for how long it takes. A few notes

* pressing 'Cancel' doesn't prevent saving but cancels the participants. 
* the configure link should bring you the places that allow to disable this behaviour


Last, I do agree save participants should be fast. You can selectively disable extensions to figure out on it is (maybe python https://github.com/microsoft/vscode/issues/209789#issuecomment-2065878927).

**Comment by camillol** [comment]
The priority inversion may be within the extension itself, not VSCode. For example, if it's Pylance, it may not be getting around to removing unused imports in the file being saved because it's busy indexing 300 other unrelated files. I'll file another issue against PyLance. For things VSCode can do:
- "Cancel" in UI almost universally means "cancel the current action as initiated by the user". How about relabeling the button to "Skip extensions" or "Skip and save"?
- VSCode seems to list every single extension that is trying to run a save action, making it hard to find the culprit. In my case it should have just said "Python" was being slow. I assume these are running sequentially and you have a timeout on the entire process, but you should also collect the timing of individual steps and only list the slow one. Don't list the ones that haven't run yet, don't list those that already ran and took a reasonable time.

**Comment by ian-h-chamberlain** [comment]
I would agree that generally yes, this is the extension's responsibility to fix performance. However, I think a reasonable ask for VSCode itself is the ability to skip some extensions when performing a save (without fully disabling that extension).

I think #82718 might be one approach to resolving this? What I'd really love for certain extensions I find generally useful but I know have slow save-actions would be some configuration like this:
```jsonc
"editor.codeActionsOnSave": {
    "source.organizeImports": "explicit",
    "source.fixAll": "explicit",
	"skip": [
		"ms-python.python",
		// list any other problematic extension IDs here...
	]
}
```

Alternatively, maybe it would be useful to have a configurable timeout to automatically cancel actions that take too long... I find myself hitting cancel after a second or two pretty often so something like 
```jsonc
"editor.codeActionsOnSave": {
	"timeoutMs": 1500
}
```
would probably also work as a quick and dirty fix.

**Comment by jrieken** [comment]
* disable auto save 
* have a fix all code action provider that is slow (see below)
* enable code actions on save
* save a JS file
* make sure you see a progress notification that 
  * mentions your extension
  * has a skip action which skips the participant but does NOT cancel saving

```
	vscode.languages.registerCodeActionsProvider('javascript', new class implements vscode.CodeActionProvider {
		async provideCodeActions(document: vscode.TextDocument, range: vscode.Range | vscode.Selection, context: vscode.CodeActionContext, token: vscode.CancellationToken): Promise<(vscode.CodeAction | vscode.Command)[]> {

			console.log('HERE, WAIT')
			await new Promise(resolve => setTimeout(() => resolve([]), 18000));

			const result = new vscode.CodeAction('Insert 42', vscode.CodeActionKind.SourceFixAll.append('foo'));
			result.edit = new vscode.WorkspaceEdit();
			result.edit.insert(document.uri, new vscode.Position(0, 0), '// 42\n')
			return [result];
		}
	}, { providedCodeActionKinds: [vscode.CodeActionKind.SourceFixAll] })
```

**Comment by jrieken** [comment]
@justschen you might be the best person to verify this ⏫

**Comment by justschen** [comment]
@jrieken will do! I'll also mention this in some of the other threads we've been seeing related to the longer save times.
